packages/timeline-state-resolver/src/service/stateHandler.ts

```diff
@@ -0,0 +1,220 @@
+			mappings: this.currentState?.mappings || {},
+		}
+		await this.calculateNextStateChange()
+	}
+
+	clearFuture(t: number) {
+		this.stateQueue = this.stateQueue.filter((s) => s.state.time <= t)
+	}
+
+	private async calculateNextStateChange() {
+		if (!this.currentState) return // a change is currently being executed, we'll be called again once it's done
+
+		const nextState = this.stateQueue[0]
+		if (!nextState) return
+
+		try {
+			const trace = startTrace('device:diffDeviceStates', { deviceId: this.context.deviceId })
+			nextState.commands = this.diffDeviceStates(
+				this.currentState.deviceState,
+				nextState.deviceState,
+				nextState.mappings
+			)
+			this.context.emitTimeTrace(endTrace(trace))
+		} catch (e) {
+			// todo - log an error
+			this.logger.error('diffDeviceState failed, t = ' + nextState.state.time, e as Error)
+			// we don't want to get stuck, so we should act as if this can be executed anyway
+			nextState.commands = []
+		}
+
+		if (nextState.state.time <= Date.now() && this.currentState) {
+			await this.executeNextStateChange()
+		}
+	}
+
+	private async executeNextStateChange() {
+		if (!this.stateQueue[0] || this._executingStateChange) {
+			// there is no next to execute - or we are currently executing something
+			return
+		}
+		this._executingStateChange = true
+
+		if (!this.stateQueue[0].commands) {
+			await this.calculateNextStateChange()
+		}
+
+		const newState = this.stateQueue.shift()
+
+		if (!newState || !newState.commands) {
+			// this should not be possible given our previous guard?
+			return
+		}
+
+		newState.measurement?.executeState()
+
+		this.currentState = undefined
+
+		if (this.config.executionType === 'salvo') {
+			Promise.allSettled(
+				newState.commands.map(async (command) => {
+					newState.measurement?.executeCommand(command)
+					return this.executeCommand(command).then(() => {
+						newState.measurement?.finishedCommandExecution(command)
+						return command
+					})
+				})
+			)
+				.then(() => {
+					if (newState.measurement) this.context.reportStateChangeMeasurement(newState.measurement.report())
+				})
+				.catch((e) => {
+					this.logger.error('Error while executing next state change', e)
+				})
+		} else {
+			const execAll = async () => {
+				for (const command of newState.commands || []) {
+					newState.measurement?.executeCommand(command)
+					await this.executeCommand(command).catch((e) => {
+						this.logger.error('Error while executing command', e)
+					})
+					newState.measurement?.finishedCommandExecution(command)
+				}
+			}
+
+			execAll()
+				.then(() => {
+					if (newState.measurement) this.context.reportStateChangeMeasurement(newState.measurement.report())
+				})
+				.catch((e) => {
+					this.logger.error('Error while executing next state change', e)
+				})
+		}
+
+		this.currentState = newState as ExecutedStateChange<DeviceState, Command>
+		this._executingStateChange = false
```

Comment:
the execution of commands is not being awaited, so this will be set while the commands are still executing. Is this intentional?

Reply:
Yes, this exists to wait for the next state to have been removed from the state queue and set as the current state. Although in my initial implementation I assumed all commands to be sent as salvos. Perhaps now that a sequential mode exists it would be better to await the last command to be sent in that mode.